File: examples/js/websocket-orderbook.js
'use strict';

const ccxt = require('../../ccxt.js')
const extend = ccxt.extend;
var fs = require('fs');

const baseConfig = {
    exchangeDefaults: {
        verbose: false
    },
    symbolDefaults: {
        limit: 5
    },
    marketTable: {
        marketsByRow: 3,
        maxLimit: 10,
        marketColumnWidth: 50,
    },
    exchanges : {
        bitfinex2 : {
            symbols: {
                "BTC/USDT": {},
                "XRP/USDT": {}
            },
            options: {

            }
        },
        binance: {
            symbols: {
                'BTC/USDT':{}
            },
            options: {

            }
        }
    }
};



var marketTable;
function main() {
    var config;
    if (process.argv.length > 2) {
        config = extend ({}, baseConfig, JSON.parse(fs.readFileSync(process.argv[2], 'utf8')));
    } else {
        config = extend ({} , baseConfig);
    }
    marketTable = new MarketTable(config.marketTable);
    for (var id in config.exchanges) {
        var exchange = config.exchanges[id];
        const exConf = extend ({}, config.exchangeDefaults, exchange.options | {});
        
        const ex = new ccxt[id]({
            apiKey: exConf.apiKey | '',
            secret: exConf.apiSecret | '',
            enableRateLimit: true,
            verbose: exConf.verbose | false,
        }); 
        for (var symbol in exchange.symbols) {
            marketTable.addMarket (ex.id, symbol);
        }
        subscribe (ex, exchange.symbols, config.symbolDefaults);
    }
}



function subscribe(exchange, symbols, defaultConfig) {
    exchange.on('err', (err, conxid) => {
        try {
          exchange.websocketClose(conxid);
        } catch (ex) {
        }
        for (var symbol in symbols) {
            marketTable.marketError(exchange.id, symbol, err);
        }
        marketTable.print();
    });
    exchange.on('ob', (market, ob) => {
        marketTable.updateMarket (exchange.id, market, ob);
        marketTable.print();
    });
    exchange.loadMarkets().
        then(async () => {
            for (var symbol in symbols) {
                const config = extend ({}, defaultConfig, symbols[symbol]);
                try {
                    await exchange.websocketSubscribe('ob', symbol, config);
                } catch (ex) {
                    marketTable.marketError(exchange.id, symbol, ex);
                    marketTable.print();
                }
            }
        }).
        catch((error) => {
            for (var symbol in symbols) {
                marketTable.marketError(exchange.id, symbol, error);
            }
            marketTable.print();
        });
}


class MarketTable {
    constructor (options){
        this.options = extend ({}, {
            marketsByRow: 2,
            maxLimit: 10,
            marketColumnWidth: 50,
        }, options);
        this.markets = {};
        this.grid = [];
        this.newEmptyLine = Array(this.options.marketColumnWidth * this.options.marketsByRow).join(" ");
        this.hr = Array(this.options.marketColumnWidth).join("-");
        this.bidsAdksSeparator = Array(this.options.marketColumnWidth).join(".");
        this.emptyCell = Array(this.options.marketColumnWidth).join(" ");
        this.amountColumn = Math.floor(this.options.marketColumnWidth / 2);
        this.height = 2+this.options.maxLimit + 1 + this.options.maxLimit + 1;
    }

    replaceAt (str, index, replacement) {
        return str.substr(0, index) + replacement + str.substr(index + replacement.length);
    }

    addMarket (marketId, marketSymbol) {
        const gridPosition = Object.keys(this.markets).length;
        const gridRow = Math.floor(gridPosition / this.options.marketsByRow) * this.height;
        const gridColumn = (gridPosition % this.options.marketsByRow) * this.options.marketColumnWidth;
        const lastUpdate = new Date();
        this.markets[marketId + '@' + marketSymbol] = {
            gridPosition,
            gridRow,
            gridColumn,
            lastUpdate,
            marketId,
            marketSymbol
        };
        // add new row
        if (this.grid.length <= gridRow) {
            const newHeight = gridRow + (this.options.maxLimit * 2) + 4;
            for (var i=this.grid.length ; i < newHeight; i++) {
                this.grid.push(this.newEmptyLine);
            }
        }
        this.grid[gridRow] = this.replaceAt (this.grid[gridRow], gridColumn, marketId + ":" + marketSymbol+ ":" + lastUpdate.toISOString());
        this.grid[gridRow+1] = this.replaceAt (this.grid[gridRow+1], gridColumn, this.hr);
        const separatorRow = gridRow+2+this.options.maxLimit;
        this.grid[separatorRow] = this.replaceAt (this.grid[separatorRow], gridColumn, this.bidsAdksSeparator);
    }

    updateMarket (marketId, marketSymbol, ob) {
        const lastUpdate =  new Date();
        const gridRow = this.markets[marketId + '@' + marketSymbol].gridRow;
        const gridColumn = this.markets[marketId + '@' + marketSymbol].gridColumn;
        this.markets[marketId + '@' + marketSymbol].lastUpdate = lastUpdate;
        // update title
        this.grid[gridRow] = this.replaceAt (this.grid[gridRow], gridColumn, marketId + ":" + marketSymbol+ ":" + lastUpdate.toISOString());
        for (var index=0,i=gridRow + 2 + this.options.maxLimit - 1; (index < this.options.maxLimit) && (index < ob.bids.length); i--, index++) {
            const price = ob.bids[index][0];
            const ammount = ob.bids[index][1];
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, this.emptyCell);
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, price.toString());
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn + this.amountColumn, ":"+ ammount);
        }
        for (; index < this.options.maxLimit; index++,i--) {
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, this.emptyCell);
        }
        for (var index=0,i = gridRow + 2 + this.options.maxLimit + 1; (index < this.options.maxLimit) && (index < ob.asks.length); i++, index++) {
            const price = ob.asks[index][0];
            const ammount = ob.asks[index][1];
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, this.emptyCell);
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, price.toString());
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn + this.amountColumn, ":"+ ammount);
        }
        for (; index < this.options.maxLimit; index++,i++) {
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, this.emptyCell);
        }
    }

    marketError (marketId, marketSymbol, error) {
        const gridRow = this.markets[marketId + '@' + marketSymbol].gridRow;
        const gridColumn = this.markets[marketId + '@' + marketSymbol].gridColumn;
        const errLines = error.toString().split(this.options.marketColumnWidth - 1);
        // this.grid[gridRow+1] = this.replaceAt (this.grid[gridRow+1], gridColumn, this.bidsAdksSeparator);
        for (var index=0,i=gridRow + 2; (index < this.options.maxLimit * 2 + 1) && (index < errLines.length); i++, index++) {
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, this.emptyCell);
            this.grid[i] = this.replaceAt (this.grid[i], gridColumn, errLines[index]);
        }
    }

    print () {
        //process.stdout.write('\\033c');
        // process.stdout.write("\u001b[2J\u001b[0;0H");
        process.stdout.write("\u001b[0;0H");
        for (var i = 0; i<this.grid.length; i++) {
            console.log(this.grid[i]);
        }
    }
}




main();

File: examples/py/websocket-orderbook.py
import os
import sys
import pprint
import traceback
import datetime

pp = pprint.PrettyPrinter(depth=6)

root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(root + '/python')
# import ccxt  # noqa: E402
import ccxt.async_support as ccxt  # noqa: E402
import asyncio  # noqa: E402
from ccxt.base.exchange import Exchange as BaseExchange

loop = asyncio.get_event_loop()

baseConfig = {
    "exchangeDefaults": {
        "verbose": False
    },
    "symbolDefaults": {
        "limit": 5
    },
    "marketTable": {
        "marketsByRow": 3,
        "maxLimit": 10,
        "marketColumnWidth": 50,
    },
    "exchanges" : {
        "bitfinex2" : {
            "symbols": {
                "BTC/USDT": {},
                "XRP/USDT": {}
            },
            "options": {

            }
        },
        "binance": {
            "symbols": {
                'BTC/USDT':{}
            },
            "options": {

            }
        }
    }
}
marketTable = None
def main():
    global marketTable
    if (len(sys.argv) > 2):
        with open(sys.argv[2]) as f:
            config = BaseExchange.extend ({}, baseConfig, json.load(f))
    else:
        config = BaseExchange.extend ({} , baseConfig)
    marketTable = MarketTable(config['marketTable'])
    for id in config['exchanges'].keys():
        exchange = config['exchanges'][id]
        exConf = BaseExchange.extend ({}, config['exchangeDefaults'], exchange['options'] if 'options' in exchange else {})
        
        ex = getattr(ccxt, id)({
            'apiKey': exConf['apiKey'] if 'apiKey' in exConf else '',
            'secret': exConf['apiSecret'] if 'apiSecret' in exConf else '',
            'enableRateLimit': True,
            'verbose': exConf['verbose'] if 'verbose' in exConf else False,
        }); 
        for symbol in exchange['symbols'].keys():
            marketTable.addMarket (ex.id, symbol)
        asyncio.ensure_future(subscribe (ex, exchange['symbols'], config['symbolDefaults']), loop=loop)
        # await subscribe (ex, exchange['symbols'], config['symbolDefaults'])

async def subscribe(exchange, symbols, defaultConfig):
    
    @exchange.on('err')
    def websocket_error(err, conxid):  # pylint: disable=W0612
        print(err)
        try:
          exchange.websocketClose(conxid)
        except ex:
            pass
        for symbol in symbols.keys():
            marketTable.marketError(exchange.id, symbol, err)
        marketTable.print()

    @exchange.on('ob')
    def websocket_ob(market, ob):  # pylint: disable=W0612
        marketTable.updateMarket (exchange.id, market, ob)
        marketTable.print()

    try:
        await exchange.load_markets()
        for symbol in symbols.keys():
            sys.stdout.flush()
            config = BaseExchange.extend ({}, defaultConfig, symbols[symbol])
            try:
                await exchange.websocket_subscribe('ob', symbol, config)
            except ex:
                print(ex)
                sys.stdout.flush()
                marketTable.marketError(exchange.id, symbol, ex)
                marketTable.print()
    except ex: 
        print(ex)
        sys.stdout.flush()
        for symbol in symbols.keys():
            marketTable.marketError(exchange.id, symbol, ex)
        marketTable.print()

class MarketTable:
    def __init__ (self, options):
        self.options = BaseExchange.extend ({}, {
            "marketsByRow": 2,
            "maxLimit": 10,
            "marketColumnWidth": 50,
        }, options)
        self.markets = {}
        self.grid = []
        self.newEmptyLine = ' ' * (self.options['marketColumnWidth'] * self.options['marketsByRow'])
        self.hr = "-" * (self.options['marketColumnWidth'])
        self.bidsAdksSeparator = "." * (self.options['marketColumnWidth'])
        self.emptyCell = " " * (self.options['marketColumnWidth'])
        self.amountColumn = self.options['marketColumnWidth'] // 2
        self.height = 2+self.options['maxLimit'] + 1 + self.options['maxLimit'] + 1

    def replaceAt (self, str, index, replacement):
        return str[:index] + replacement + str[index + len(replacement):]

    def addMarket (self, marketId, marketSymbol):
        gridPosition = len(self.markets.keys())
        gridRow = (gridPosition // self.options['marketsByRow']) * self.height
        gridColumn = (gridPosition % self.options['marketsByRow']) * self.options['marketColumnWidth']
        lastUpdate = datetime.datetime.now()
        self.markets[marketId + '@' + marketSymbol] = {
            'gridPosition': gridPosition,
            'gridRow': gridRow,
            'gridColumn': gridColumn,
            'lastUpdate': lastUpdate,
            'marketId': marketId,
            'marketSymbol': marketSymbol
        }
        # add new row
        if (len(self.grid) <= gridRow):
            newHeight = gridRow + (self.options['maxLimit'] * 2) + 4
            for i in range(len(self.grid), newHeight):
                self.grid.append(self.newEmptyLine)

        self.grid[gridRow] = self.replaceAt (self.grid[gridRow], gridColumn, marketId + ":" + marketSymbol+ ":" + str(lastUpdate))
        self.grid[gridRow+1] = self.replaceAt (self.grid[gridRow+1], gridColumn, self.hr)
        separatorRow = gridRow+2+self.options['maxLimit']
        self.grid[separatorRow] = self.replaceAt (self.grid[separatorRow], gridColumn, self.bidsAdksSeparator)

    def updateMarket (self, marketId, marketSymbol, ob):
        lastUpdate = datetime.datetime.now()
        gridRow = self.markets[marketId + '@' + marketSymbol]['gridRow']
        gridColumn = self.markets[marketId + '@' + marketSymbol]['gridColumn']
        self.markets[marketId + '@' + marketSymbol]['lastUpdate'] = lastUpdate
        # update title
        self.grid[gridRow] = self.replaceAt (self.grid[gridRow], gridColumn, marketId + ":" + marketSymbol+ ":" + str(lastUpdate))
        i = gridRow + 2 + self.options['maxLimit'] - 1
        for index in range (0,min(self.options['maxLimit'],len(ob['bids']))):
            price = ob['bids'][index][0]
            ammount = ob['bids'][index][1]
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, self.emptyCell)
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, str(price))
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn + self.amountColumn, ":"+ str(ammount))
            i = i - 1
        for index in range(index + 1, self.options['maxLimit']):
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, self.emptyCell)
            i = i - 1
        i = gridRow + 2 + self.options['maxLimit'] + 1
        for index in range (0,min(self.options['maxLimit'],len(ob['asks']))):
            price = ob['asks'][index][0]
            ammount = ob['asks'][index][1]
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, self.emptyCell)
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, str(price))
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn + self.amountColumn, ":"+ str(ammount))
            i = i + 1
        for index in range(index + 1, self.options['maxLimit']):
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, self.emptyCell)
            i = i + 1

    def marketError (self, marketId, marketSymbol, error):
        gridRow = self.markets[marketId + '@' + marketSymbol]['gridRow']
        gridColumn = self.markets[marketId + '@' + marketSymbol]['gridColumn']
        errText = str(error)
        errLines = [errText[i:i+n] for i in range(0, len(errText), self.options['marketColumnWidth'] - 1)]
        for index in range (0,min(self.options['maxLimit'],len(errLines))):
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, self.emptyCell)
            self.grid[i] = self.replaceAt (self.grid[i], gridColumn, errLines[index])

    def print (self):
        sys.stdout.write("\u001b[0;0H")
        for line in self.grid:
            print(line)
        sys.stdout.flush()


#loop.run_until_complete(main())
main()
loop.run_forever()